### ISSUE TITLE
[Gemma 4 E2B] `ttnn.embedding` gives incorrect results if `hidden_dim=8960`

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-metal#44500]
### Component / Area

ops

### Issue Type (optional)

Bad Outputs

### Observed

For `hidden_dim=8960` , as it is in Gemma 4 E2B, `ttnn.embedding` gives incorrect results.
In repro, `seq_len = 1024`, as it is actual use-case, however same op gives incorrect results even if `seq_len = 32`.

### Expected

`ttnn.embedding` should give correct results.

### 1. Steps (exact commands)

Run
```python
import ttnn, torch

VOCAB = 262144
HIDDEN = 8960       
SEQ = 1024

with ttnn.manage_device(device_id=0) as device:
    weight_torch = 0.01 * torch.randn((VOCAB, HIDDEN), dtype=torch.float32)
    ids_torch = torch.full((1, SEQ), 1, dtype=torch.int32)

    weight = ttnn.from_torch(
        weight_torch.to(torch.bfloat16),
        layout=ttnn.TILE_LAYOUT,
        device=device,
        dtype=ttnn.bfloat16,
    )
    ids = ttnn.from_torch(
        ids_torch,
        layout=ttnn.ROW_MAJOR_LAYOUT,
        device=device,
        dtype=ttnn.uint32,
    )

    out_tt = ttnn.to_torch(ttnn.embedding(ids, weight)).to(torch.float32)

    weight_back = ttnn.to_torch(weight).to(torch.float32)
    out_torch = torch.nn.functional.embedding(ids_torch.long(), weight_back)

    diff = (out_tt - out_torch).abs()
    print("max abs diff:", diff.max().item())

    # Last 16 elements of out_tt and out_torch should be the same
    print("last 16 elements of out_tt:", out_tt[-16:])
    print("last 16 elements of out_torch:", out_torch[-16:])

    # Number of infs in out_tt and out_torch
    print("number of infs in out_tt:", (out_tt == float('inf')).sum().item())
    print("number of infs in out_torch:", (out_torch == float('inf')).sum().item())
```

### 2. Input data / link or description

none

### 3. Frequency

Always

### 1. Software Versions

commit d5a16537336229bee54fd4a6d8bd54c492abc7d1

### 2. Hardware Details

Blackhole p150

### Is this a regression?

Unknown

### Logs & Diagnostics

Output of above repro:
```
max abs diff: inf
last 16 elements of out_tt: tensor([[[-7.8964e-04,  9.4604e-03, -3.6774e-03,  ...,  1.1396e-04,
           2.2736e-03,  1.0071e-02],
         [-8.9111e-03, -1.9684e-03, -1.8066e-02,  ...,  2.1118e-02,
           1.9684e-03,  5.7373e-03],
         [ 4.9973e-04, -2.1606e-02, -3.3264e-03,  ..., -1.7334e-02,
           1.2756e-02, -1.1475e-02],
         ...,
         [ 0.0000e+00,  3.0078e-01,  0.0000e+00,  ..., -1.2146e-02,
           0.0000e+00, -9.3384e-03],
         [ 0.0000e+00,  5.8838e-02,  0.0000e+00,  ..., -3.1738e-02,
           0.0000e+00,  1.0645e-01],
         [ 0.0000e+00,  3.1982e-02,  0.0000e+00,  ...,  7.0801e-03,
           0.0000e+00, -4.0588e-03]]])
last 16 elements of out_torch: tensor([[[-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053],
         [-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053],
         [-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053],
         ...,
         [-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053],
         [-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053],
         [-0.0008,  0.0095, -0.0037,  ..., -0.0017,  0.0031,  0.0053]]])
number of infs in out_tt: 580
number of infs in out_torch: 0
```

### Priority

None

### Impact

Affects bringup of Gemma 4 E2B through tt-forge.

**Comment by yieldthought** [comment]
AutoDebug could not finish this request.

Request from @yieldthought: (no extra request text)

Issue: https://github.com/tenstorrent/tt-metal/issues/44500

Error: `autodebug did not complete successfully.`

**Comment by yieldthought** [comment]
@yieldthought autodebug

**Comment by yieldthought** [comment]
✅ AutoDebug top finding: Chunked ttnn.embedding mishandles non-64-multiple widths like 8960, causing stale L1 reads and corrupted output.

<details>
<summary>Full report</summary>

# AutoDebug Report: tt-metal issue 44500

## Summary

`ttnn.embedding` is selecting the fused row-major-indices to tiled-output implementation for the reported repro, then switching to its chunked tilize path because `hidden_dim=8960` needs more than the hardcoded 1 MiB L1 budget. In that chunked path the reader splits the embedding row by `num_chunks` bytes, while the compute kernel and circular-buffer protocol still treat every chunk as `tiles_per_chunk` full tiles.

For `hidden_dim=8960`, this means the reader writes 56 tile-columns per token row but advertises 64 tile-columns to compute. The compute kernel then tilizes with the wrong row stride, consumes unwritten/stale L1 as valid data, and produces 320 tiles per 32-token block while the writer only writes 280. This is a deterministic code-level mismatch and is sufficient to explain the bad values and `inf`s in the reported output.

I did not run a hardware repro.

## Headline Finding: chunked fused embedding corrupts widths like 8960

### How the repro reaches this path

The Python repro calls `ttnn.embedding(ids, weight)` with row-major `uint32` ids and a tiled `bfloat16` weight. The C++ wrapper converts tiled weights to row-major before launching the primitive op, but it still uses the original weight layout to decide default tiled output:

- `embedding.cpp` converts `weight_arg` from tile to row-major before the primitive op: `ttnn/cpp/ttnn/operations/embedding/embedding.cpp:28-35`.
- Row-major ids are reshaped to `{batch, 1, 1, sentence_size}`: `ttnn/cpp/ttnn/operations/embedding/embedding.cpp:40-43`.
- Since `seq_len=1024` and `hidden_dim=8960` are both tile multiples, and no explicit layout was passed, the original tiled weight makes `fused_tilized = true`: `ttnn/cpp/ttnn/operations/embedding/embedding.cpp:45-55`.
- Row-major input plus `tilized=true` selects `EmbeddingsFusedProgramFactory`: `ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:17-25`.
- The primitive output shape is `{1, 1, 1024, 8960}` in tile layout: `ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:110-126`.

### Concrete failing arithmetic

Constants:

- Tile width/height are 32: `tt_metal/api/tt-metalium/constants.hpp:13-14`.
- A bf16 tile is 2048 bytes: `tt_metal/api/tt-metalium/tt_backend_api_types.hpp:109-110`.

For `hidden_dim=8960`:

- `num_tiles_per_block = 8960 / 32 = 280`: `ttnn/cpp/ttnn/operations/embedding/device/embeddings_fused_program_factory.cpp:63-74`.
- Full double-buffered embedding block would need `2 * 280 * 2048 = 1,146,880` bytes.
- The factory enables chunking when this exceeds `1024 * 1024`: `ttnn/cpp/ttnn/operations/embedding/device/embeddings_fused_program_factory.cpp:93-101`.
- It chooses `tiles_per_chunk = 64` and `num_chunks = ceil(280 / 64) = 5`: `ttnn/cpp/ttnn/operations/embedding/device/embeddings_fused_program_factory.cpp:108-117`.

The bug is in the reader/compute contract:

- The reader computes `weight_chunk_size = weight_block_size / num_chunks`: `ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embeddings_tilize.cpp:49-51`.
- Here `weight_block_size = 8960 * 2 = 17,920`, so `weight_chunk_size = 17,920 / 5 = 3,584` bytes.
- `3,584` bytes is only `3,584 / (32 * 2) = 56` tile-columns per token row.
- But the reader reserves and pushes `tiles_per_chunk = 64` tile pages: `ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embeddings_tilize.cpp:45-61`.
- The compute kernel instantiates `tilize<tiles_per_chunk>` and runs `per_core_block_cnt * num_chunks`, so it consumes and produces `64 * 5 = 320` tiles per 32-token block: `ttnn/cpp/ttnn/operations/embedding/device/kernels/compute/tilize_chunked.cpp:10-27`.
- The DRAM writer is only told to write the true block width, `num_tiles_per_block * local_num_blocks = 280 * local_num_blocks`: `ttnn/cpp/ttnn/operations/embedding/device/embeddings_fused_program_factory.cpp:274-280`.

This is not just a harmless tail overproduction. For each chunk, the reader writes each token row compactly with a 3,584-byte stride, but `tilize<64>` interprets the chunk as a 64-tile-wide row-major block, which requires a 4,096-byte stride. Starting with row 1 in the 32-token tile block, compute reads from the wrong offset; near the end of each chunk it also reads pages that the reader never wrote but still pushed as valid. That explains why all token ids can be `1` while output rows differ, and why stale data can show up as `inf`.

### Why the reported boundaries make sense

- `seq_len=32` can still fail because one 32-token block is enough to trigger the bad per-block chunk contract. Larger sequence lengths only increase the number of affected blocks.
- `hidden_dim=8192` stays on the non-chunked path: `8192 / 32 = 256`, and `2 * 256 * 2048 = 1,048,576`, exactly the budget. The code checks `>`, not `>=`.
- The existing large-width unit test uses `hidden_embedding_dim=16384`: `tests/ttnn/unit_tests/operations/data_movement/test_embedding.py:517-543`. That is `512` tiles, which divides evenly into eight 64-tile chunks, so the current byte-splitting bug is masked.
- A width like `10240` also happens to divide into five 64-tile chunks. `8960` does not: it is 280 tiles, so five chunks should not all be 64 real tiles.

### Suggested fix direction

The chunked reader/compute contract needs an explicit partial-tail concept. The direct fix is to pass `last_chunk_tiles = num_tiles_per_block - (num_chunks - 1) * tiles_per_chunk`, have the reader push `last_chunk_tiles` pages for the final chunk, and have the compute kernel tilize the final chunk with `tilize<last_chunk_tiles>` instead of treating every chunk as `tiles_per_chunk`.

There is already a local remote branch with that shape of fix: `origin/sshon/pr-embedding-pcc-issue`, commit `d81057b26ee` (`ttnn.embedding: fix partial trailing chunk in chunked tilize path`). It adds `last_chunk_tiles` to the fused factory, reader, and compute kernel, and adds regression tests for non-64-multiple chunked widths such as `8224`, `8320`, `10752`, and `15488`. That branch is not the checked-out issue commit (`d5a16537336229bee54fd4a6d8bd54c492abc7d1`), but its diff matches the root-cause arithmetic above.

Regression tests should include a chunked width that is not a multiple of 64 tiles, for example:

- `hidden_embedding_dim=8960`, `seq_len=32`, all ids equal, tiled output.
- A near-threshold case such as `hidden_embedding_dim=8224` (`257` tiles), which would also catch non-tile-aligned/truncated byte division.
- A control case such as `hidden_embedding_dim=10240` or `16384`, which is chunked but evenly divisible by 64 tiles.

## Other Potential Issues / Test Gaps

The nanobind doc says the `layout` keyword defaults to `ttnn.ROW_MAJOR_LAYOUT`, but the binding default is `None` and the wrapper ultimately uses `layout.value_or(weight_arg.layout())`: `ttnn/cpp/ttnn/operations/embedding/embedding_nanobind.cpp:28-38`, `ttnn/cpp/ttnn/operations/embedding/embedding_nanobind.cpp:47-52`, and `ttnn/cpp/ttnn/operations/embedding/embedding.cpp:75-78`. This documentation mismatch is not the root cause, but it makes the repro's default tiled output less obvious.

The sweep coverage does not exercise this failing fused tiled-output path at large non-64-multiple widths. The sweep shape generator tops out at `embedding_dim=4128`, and invalidation excludes tiled weights: `tests/sweep_framework/sweeps/embedding/embedding.py:27-37` and `tests/sweep_framework/sweeps/embedding/embedding.py:42-51`. The regular unit tests cover tiled output, but only up to `4096` except for the masked `16384` OOM regression case.

There is a separate tiled-index/BFP16-index code discrepancy in `embedding_ind_tilized.cpp`: `read_block` first reads `input_l1_ptr[token_idx + offset]`, but the BFP16 branch ignores that value and recasts `input_l1_ptr[token_idx]` without the offset (`ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:42-72`). This is not on the reported path because the repro uses row-major `uint32` ids, but it is worth auditing separately.

## Remaining Uncertainty

Without Blackhole hardware I cannot prove the exact runtime manifestation, timing, or final tensor contents. The code-level mismatch is deterministic for the reported dimensions, though: the reader pushes a different number of valid tile pages than the compute kernel consumes, with a row stride that contradicts the compiled tilize width.

## Continue In Codex

To reopen this investigation thread:

```bash
codex resume 019e4ff4-d8cf-7d42-8efb-782cdb737cdb
```

</details>
